src/aspire/orientation/commonline_sync.py

```diff
@@ -1,34 +1,298 @@
 from aspire.orientation import Orient3D
 import logging
+import numpy as np
+
+import scipy.sparse as sps
+import scipy.sparse.linalg as spsl
 
 logger = logging.getLogger(__name__)
 
 
 class CommLineSync(Orient3D):
     """
-    Define a class to estimate 3D orientations using Synchronization described as below:
-    Y. Shkolnisky, and A. Singer, Viewing Direction Estimation in Cryo-EM Using Synchronization,
+    Define a class to estimate 3D orientations using synchronization matrix and voting method.
+
+    The related publications are listed as below:
+    Y. Shkolnisky, and A. Singer,
+    Viewing Direction Estimation in Cryo-EM Using Synchronization,
     SIAM J. Imaging Sciences, 5, 1088-1110 (2012).
 
+    A. Singer, R. R. Coifman, F. J. Sigworth, D. W. Chester, Y. Shkolnisky,
+    Detecting Consistent Common Lines in Cryo-EM by Voting,
+    Journal of Structural Biology, 169, 312-322 (2010).
     """
-    def __init__(self, src):
+    def __init__(self, src, nrad=None, ntheta=None):
+        """
+        Initialize an object for estimating 3D orientations using synchronization matrix
+
+        :param src: The source object of 2D denoised or class-averaged images with metadata
+        :param n_rad: The number of points in the radial direction
+        :param n_theta: The number of points in the theta direction
+        """
+        super().__init__(src, nrad=nrad, ntheta=ntheta)
+        self.syncmatrix = None
+
+    def estimate_rotations(self):
+        """
+        Estimate orientation matrices for all 2D images using synchronization matrix
+        """
+        if self.syncmatrix is None:
+            self.syncmatrix_vote()
+
+        s = self.syncmatrix
+        sz = s.shape
+        if len(sz) != 2:
+            raise ValueError('clmatrix must be a square matrix')
+        if sz[0] != sz[1]:
+            raise ValueError('clmatrix must be a square matrix')
+        if sz[0] % 2 == 1:
+            raise ValueError('clmatrix must be a square matrix of size 2Kx2K')
+
+        k = sz[0] // 2
+
+        d, v = sps.linalg.eigs(s, 10)
+        d = np.real(d)
+        sort_idx = np.argsort(-d)
+
+        v = np.real(v[:, sort_idx[:3]])
+        v1 = v[:2*k:2].T.copy()
+        v2 = v[1:2*k:2].T.copy()
+
+        equations = np.zeros((3*k, 9))
+        for i in range(3):
+            for j in range(3):
+                equations[0::3, 3*i+j] = v1[i] * v1[j]
+                equations[1::3, 3*i+j] = v2[i] * v2[j]
+                equations[2::3, 3*i+j] = v1[i] * v2[j]
+        truncated_equations = equations[:, [0, 1, 2, 4, 5, 8]]
+
+        b = np.ones(3 * k)
+        b[2::3] = 0
+
+        ata_vec = np.linalg.lstsq(truncated_equations, b, rcond=None)[0]
+        ata = np.zeros((3, 3))
+        ata[0, 0] = ata_vec[0]
```

Comment:
Again, this is very mysterious…

Reply:
These statements try to find the least squares approximation of `A'*A` in vector form and convert back to 3x3 matrix. Added more explanatory comments for better understanding in commit 614aaa8.

Reply:
Ah ok. Might be more explicit to write this as
```
upper_mask = np.triu(np.full((3, 3), True))
ata[upper_mask] = ata_vec

ata = (ata + ata.T) / 2
```

Reply:
The averaging can not be applied to all elements except diagonal ones. Done by         
```suggestion
        upper_mask = np.triu(np.full((3, 3), True))
        ATA[upper_mask] = ATA_vec
        lower_mask = np.tril(np.full((3, 3), True))
        ATA[lower_mask] = ATA.T[lower_mask]
```

Reply:
I don't understand. Why can't the averaging be applied?

Reply:
The Non diagonal elements from upper_mask and lower mask  were calculated once only. Averaging will reduce them by a factor of 0.5.

Reply:
Ah ok.